File: scripts/build-search-index.mjs
import fs from 'fs';
import path from 'path';

const CWD = process.cwd();

async function generateSearchIndex() {
  const searchIndex = [];

  // 1. Index intro.md
  try {
    const introContent = fs.readFileSync(path.join(CWD, 'intro.md'), 'utf-8');
    searchIndex.push({
      id: 'intro',
      title: '项目介绍与目录结构',
      path: '/',
      content: introContent,
    });
    console.log('Indexed intro.md');
  } catch (error) {
    console.error('Error indexing intro.md:', error);
  }

  // 2. Index projects from data/projects.ts
  try {
    const projectsTSContent = fs.readFileSync(path.join(CWD, 'data/projects.ts'), 'utf-8');
    const projectsJSONString = projectsTSContent
      .replace(/^export const projects = /, '')
      .replace(/;\s*$/, '') // More robustly remove trailing semicolon and whitespace
      .trim();
    
    // Use new Function to parse the JS array string. Add parentheses for safety.
    const projectsData = new Function("return (" + projectsJSONString + ");")();

    projectsData.forEach(project => {
      searchIndex.push({
        id: `project-${project.slug}`,
        title: project.title,
        path: `/projects/${project.slug}`,
        content: `${project.title} ${project.brief} ${(project.tags || []).join(' ')} ${project.background || ''} ${(project.contributions || []).join(' ')} ${(project.outcomes || []).join(' ')}`,
        tags: project.tags || []
      });
    });
    console.log('Indexed ' + projectsData.length + ' projects');
  } catch (error) {
    console.error('Error indexing projects from data/projects.ts:', error);
    console.log("Consider converting data/projects.ts to a .js file and using module.exports, or ensure your Node environment can import .ts files directly (e.g., using ts-node or esbuild-register).")
  }

  // 3. Index tools from content/tools/
  try {
    const toolsDir = path.join(CWD, 'content/tools');
    if (fs.existsSync(toolsDir)) {
      // Index main tools page
      searchIndex.push({
        id: 'tools',
        title: '开发工具',
        path: '/tools',
        content: '我的开源开发工具集合，涵盖了 macOS 自动化、编辑器配置、命令行环境等各个方面，旨在提升开发效率和体验。'
      });

      // Read individual tool files
      const toolFiles = ['execute.md', 'neovim.md', 'zsh.md'];
      toolFiles.forEach(filename => {
        try {
          const toolPath = path.join(toolsDir, filename);
          if (fs.existsSync(toolPath)) {
            const toolContent = fs.readFileSync(toolPath, 'utf-8');
            const toolName = filename.replace('.md', '');
            
            // Extract title from first heading
            const titleMatch = toolContent.match(/^#\s+(.+)/m);
            const title = titleMatch ? titleMatch[1] : `${toolName} 工具`;
            
            // Get first few paragraphs as content
            const contentMatch = toolContent.match(/\n\n(.+?)(?:\n\n|$)/);
            const description = contentMatch ? contentMatch[1] : '';
            
            searchIndex.push({
              id: `tool-${toolName}`,
              title: title,
              path: `/tools#${toolName}`,
              content: `${title} ${description} ${toolContent.substring(0, 500)}`,
              tags: ['开发工具', '开源', toolName]
            });
          }
        } catch (error) {
          console.error(`Error indexing tool ${filename}:`, error);
        }
      });
      console.log('Indexed tools from content/tools/');
    }
  } catch (error) {
    console.error('Error indexing tools:', error);
  }

  // 4. Index static pages (manual entries)
  const staticPages = [
    { id: 'about', title: '关于我', path: '/about', content: '关于我的信息，技术栈，经验等。' },
    { id: 'contact', title: '联系方式', path: '/contact', content: '如何联系我，邮箱，社交媒体等。' },
    { id: 'projects', title: '项目列表', path: '/projects', content: '我的项目经验列表展示。' },
    { id: 'research', title: '研究方向', path: '/research', content: '我的研究方向和兴趣领域。' },
  ];

  staticPages.forEach(page => {
    searchIndex.push(page);
  });
  console.log('Indexed ' + staticPages.length + ' static pages');

  // 5. Write to public/search-index.json
  try {
    if (!fs.existsSync(path.join(CWD, 'public'))) {
      fs.mkdirSync(path.join(CWD, 'public'), { recursive: true });
    }
    fs.writeFileSync(
      path.join(CWD, 'public/search-index.json'),
      JSON.stringify(searchIndex, null, 2),
      'utf-8'
    );
    console.log('Search index successfully generated at public/search-index.json');
  } catch (error) {
    console.error('Error writing search-index.json:', error);
  }
}

generateSearchIndex(); 

File: scripts/build-search-index.py
#!/opt/homebrew/bin/python3
"""Build FTS5 search index from content/ markdown files."""

import os
import re
import sqlite3
from pathlib import Path

CONTENT_DIR = Path(__file__).resolve().parent.parent / "content"
DB_PATH = Path(__file__).resolve().parent.parent / "data" / "search.db"

CATEGORY_MAP = {
    "blog": "博客",
    "projects": "项目",
    "research": "研究成果",
    "tools": "工具",
    "resume": "简历",
    "about": "关于",
}

URL_MAP = {
    "blog": "/blog",
    "projects": "/projects",
    "research": "/research",
    "tools": "/tools",
    "resume": "/resume",
    "about": "/about",
}


def parse_frontmatter(text: str) -> tuple[dict, str]:
    """Parse YAML frontmatter between --- delimiters. Returns (metadata, body)."""
    if not text.startswith("---"):
        return {}, text

    parts = text.split("---", 2)
    if len(parts) < 3:
        return {}, text

    meta = {}
    for line in parts[1].strip().splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        # Simple key: value parsing (handles quoted and unquoted values)
        match = re.match(r'^(\w[\w-]*)\s*:\s*(.+)$', line)
        if match:
            key = match.group(1)
            value = match.group(2).strip()
            # Strip quotes
            if (value.startswith('"') and value.endswith('"')) or \
               (value.startswith("'") and value.endswith("'")):
                value = value[1:-1]
            # Handle YAML arrays like ["tag1", "tag2"]
            if value.startswith("[") and value.endswith("]"):
                items = re.findall(r'"([^"]*)"', value)
                if not items:
                    items = re.findall(r"'([^']*)'", value)
                value = ", ".join(items)
            meta[key] = value

    body = parts[2].strip()
    return meta, body


def strip_markdown(text: str) -> str:
    """Remove markdown syntax from text."""
    # Remove images
    text = re.sub(r'!\[([^\]]*)\]\([^)]+\)', r'\1', text)
    # Remove links, keep text
    text = re.sub(r'\[([^\]]+)\]\([^)]+\)', r'\1', text)
    # Remove HTML tags
    text = re.sub(r'<[^>]+>', '', text)
    # Remove code blocks
    text = re.sub(r'```[\s\S]*?```', '', text)
    # Remove inline code
    text = re.sub(r'`[^`]+`', '', text)
    # Remove headers markers
    text = re.sub(r'^#{1,6}\s+', '', text, flags=re.MULTILINE)
    # Remove bold/italic
    text = re.sub(r'\*{1,3}([^*]+)\*{1,3}', r'\1', text)
    # Remove horizontal rules
    text = re.sub(r'^---+$', '', text, flags=re.MULTILINE)
    # Remove blockquotes
    text = re.sub(r'^>\s+', '', text, flags=re.MULTILINE)
    # Remove badge images (shields.io etc)
    text = re.sub(r'\[!\[.*?\]\(.*?\)\]\(.*?\)', '', text)
    # Collapse whitespace
    text = re.sub(r'\n{3,}', '\n\n', text)
    return text.strip()


def get_category(filepath: Path) -> str | None:
    """Determine category from file path relative to content dir."""
    rel = filepath.relative_to(CONTENT_DIR)
    top_dir = rel.parts[0]
    return top_dir if top_dir in CATEGORY_MAP else None


def get_slug(filepath: Path) -> str:
    """Extract slug from filename."""
    return filepath.stem


def build_url(category: str, slug: str, filepath: Path) -> str:
    """Build URL path from category and slug."""
    base = URL_MAP.get(category, f"/{category}")

    # _index files map to the category root
    if slug == "_index":
        return base

    # projects/items/xxx -> /projects/xxx
    if category == "projects":
        return f"{base}/{slug}"

    # blog posts get their own page
    if category == "blog":
        return f"{base}/{slug}"

    # resume versions
    if category == "resume" and slug != "_index":
        return f"{base}/{slug}"

    # tools get their own page
    if category == "tools" and slug != "_index":
        return f"{base}"  # tools page is single

    # about/research are single pages
    return base


def main():
    DB_PATH.parent.mkdir(parents=True, exist_ok=True)

    db = sqlite3.connect(str(DB_PATH))
    db.execute("DROP TABLE IF EXISTS search_index")
    db.execute("""
        CREATE VIRTUAL TABLE IF NOT EXISTS search_index USING fts5(
            slug, url, title, description, category, tags, content,
            tokenize='unicode61'
        )
    """)

    count = 0
    for md_file in sorted(CONTENT_DIR.rglob("*.md")):
        # Skip non-content directories
        category = get_category(md_file)
        if category is None:
            continue

        # Skip home and global content
        if category in ("home", "global", "contact", "project-source", "resume-source"):
            continue

        raw = md_file.read_text(encoding="utf-8")
        meta, body = parse_frontmatter(raw)

        title = meta.get("title", "")
        description = meta.get("description", meta.get("excerpt", meta.get("brief", "")))
        tags = meta.get("tags", "")
        slug = get_slug(md_file)
        url = build_url(category, slug, md_file)
        category_label = CATEGORY_MAP.get(category, category)
        content = strip_markdown(body)

        # Skip if no meaningful content
        if not title and not content:
            continue

        db.execute(
            "INSERT INTO search_index (slug, url, title, description, category, tags, content) VALUES (?, ?, ?, ?, ?, ?, ?)",
            (slug, url, title, description, category_label, tags, content),
        )
        count += 1

    db.commit()
    db.close()
    print(f"Indexed {count} documents into {DB_PATH}")


if __name__ == "__main__":
    main()
